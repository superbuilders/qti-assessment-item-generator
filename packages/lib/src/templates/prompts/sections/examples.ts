import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
