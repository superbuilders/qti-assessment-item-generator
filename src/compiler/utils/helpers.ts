import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
