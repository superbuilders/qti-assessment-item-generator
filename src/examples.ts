import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
