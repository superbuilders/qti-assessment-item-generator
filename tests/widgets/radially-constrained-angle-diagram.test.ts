import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { radiallyConstrainedAngleDiagramExamples } from "../../examples/radially-constrained-angle-diagram"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidgetForTest } from "../helpers/generateWidgetForTest"

describe("Widget: radially-constrained-angle-diagram", () => {
	const examples: WidgetInput[] = radiallyConstrainedAngleDiagramExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidgetForTest(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})

	test("throws error for invalid ray label reference", async () => {
		const result = await errors.try(
			generateWidgetForTest({
				type: "radiallyConstrainedAngleDiagram",
				width: 400,
				height: 400,
				centerLabel: "O",
				rayLabels: ["A", "B"],
				angles: [{ fromRayLabel: "A", toRayLabel: "C", value: 60, color: "#ff6b6b" }]
			})
		)
		expect(result.error).toBeTruthy()
	})

	test("throws error for angles exceeding 360 degrees", async () => {
		const result = await errors.try(
			generateWidgetForTest({
				type: "radiallyConstrainedAngleDiagram",
				width: 400,
				height: 400,
				centerLabel: "O",
				rayLabels: ["A", "B", "C"],
				angles: [
					{ fromRayLabel: "A", toRayLabel: "B", value: 200, color: "#ff6b6b" },
					{ fromRayLabel: "B", toRayLabel: "C", value: 200, color: "#4ecdc4" }
				]
			})
		)
		expect(result.error).toBeTruthy()
	})

	test("throws error for invalid ray order", async () => {
		const result = await errors.try(
			generateWidgetForTest({
				type: "radiallyConstrainedAngleDiagram",
				width: 400,
				height: 400,
				centerLabel: "O",
				rayLabels: ["A", "B", "C"],
				angles: [{ fromRayLabel: "C", toRayLabel: "A", value: 60, color: "#ff6b6b" }]
			})
		)
		expect(result.error).toBeTruthy()
	})
})
