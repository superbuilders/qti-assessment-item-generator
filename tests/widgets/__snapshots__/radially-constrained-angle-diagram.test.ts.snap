// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: radially-constrained-angle-diagram should produce consistent output for example #1 1`] = `"<svg width="456" height="440" viewBox="-28 -20 456 440" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 200 200 L 140 200 A 60 60 0 0 1 170 148.03847577293368 Z" fill="#ff6b6b" stroke="none"/><path d="M 200 200 L 170 148.03847577293368 A 60 60 0 0 1 260 200 Z" fill="#4ecdc4" stroke="none"/><line x1="200" y1="200" x2="20" y2="199.99999999999997" stroke="#000000" stroke-width="2"/><circle cx="20" cy="199.99999999999997" r="4" fill="#000000"/><text x="-2" y="199.99999999999997" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><line x1="200" y1="200" x2="110.00000000000003" y2="44.11542731880104" stroke="#000000" stroke-width="2"/><circle cx="110.00000000000003" cy="44.11542731880104" r="4" fill="#000000"/><text x="99.00000000000003" y="25.06286843554338" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><line x1="200" y1="200" x2="380" y2="200" stroke="#000000" stroke-width="2"/><circle cx="380" cy="200" r="4" fill="#000000"/><text x="402" y="200" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><text x="123.78976446696942" y="155.99999999999997" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">60°</text><text x="244" y="123.7897644669694" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">120°</text><circle cx="200" cy="200" r="6" fill="#000000"/><text x="200" y="225" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">O</text></svg>"`;

exports[`Widget: radially-constrained-angle-diagram should produce consistent output for example #2 1`] = `"<svg width="575" height="562" viewBox="-38 -20 575 562" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 250 250 L 243.02754058018735 329.6955758473396 A 80 80 0 0 1 170.30442415266037 243.02754058018732 Z" fill="#e74c3c" stroke="none"/><path d="M 250 250 L 170.30442415266037 243.02754058018732 A 80 80 0 0 1 329.6955758473396 243.02754058018735 Z" fill="#3498db" stroke="none"/><path d="M 250 250 L 329.6955758473396 243.02754058018735 A 80 80 0 0 1 256.97245941981265 329.6955758473396 Z" fill="#2ecc71" stroke="none"/><line x1="250" y1="250" x2="229.08262174056202" y2="489.08672754201893" stroke="#000000" stroke-width="2"/><circle cx="229.08262174056202" cy="489.08672754201893" r="4" fill="#000000"/><text x="227.16519540011353" y="511.00301090003734" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">P</text><line x1="250" y1="250" x2="10.913272457981066" y2="229.08262174056193" stroke="#000000" stroke-width="2"/><circle cx="10.913272457981066" cy="229.08262174056193" r="4" fill="#000000"/><text x="-11.003010900037339" y="227.16519540011345" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">Q</text><line x1="250" y1="250" x2="489.08672754201893" y2="229.08262174056205" stroke="#000000" stroke-width="2"/><circle cx="489.08672754201893" cy="229.08262174056205" r="4" fill="#000000"/><text x="511.00301090003734" y="227.16519540011356" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">R</text><line x1="250" y1="250" x2="270.91737825943795" y2="489.08672754201893" stroke="#000000" stroke-width="2"/><circle cx="270.91737825943795" cy="489.08672754201893" r="4" fill="#000000"/><text x="272.8348045998864" y="511.00301090003734" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">S</text><text x="167.26720014315038" y="319.42106184614624" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">90°</text><text x="250.00000000000003" y="142" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">170°</text><text x="332.7327998568496" y="319.42106184614624" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">90°</text><circle cx="250" cy="250" r="6" fill="#000000"/></svg>"`;

exports[`Widget: radially-constrained-angle-diagram should produce consistent output for example #3 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 150 150 L 134.6926627053964 113.04481869954853 A 40 40 0 0 1 165.3073372946036 113.04481869954853 Z" fill="#9b59b6" stroke="none"/><line x1="150" y1="150" x2="104.07798811618923" y2="39.134456098645586" stroke="#000000" stroke-width="2"/><circle cx="104.07798811618923" cy="39.134456098645586" r="4" fill="#000000"/><text x="95.65895260415726" y="18.80910638339728" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">M</text><line x1="150" y1="150" x2="195.9220118838108" y2="39.134456098645586" stroke="#000000" stroke-width="2"/><circle cx="195.9220118838108" cy="39.134456098645586" r="4" fill="#000000"/><text x="204.34104739584276" y="18.80910638339728" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">N</text><text x="150" y="82" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">45°</text><circle cx="150" cy="150" r="6" fill="#000000"/><text x="150" y="175" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">X</text></svg>"`;

exports[`Widget: radially-constrained-angle-diagram should produce consistent output for example #4 1`] = `"<svg width="540" height="562" viewBox="-20 -42 540 562" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 250 250 L 177.495377037068 283.8094609392559 A 80 80 0 0 1 176.3596117238048 218.74150972085806 Z" fill="#4A90E2" stroke="none"/><path d="M 250 250 L 176.3596117238048 218.74150972085806 A 80 80 0 0 1 256.97245941981265 170.30442415266037 Z" fill="#F5A623" stroke="none"/><path d="M 250 250 L 256.97245941981265 170.30442415266037 A 80 80 0 0 1 324.174708365343 220.03147252672704 Z" fill="#7ED321" stroke="none"/><path d="M 250 250 L 324.174708365343 220.03147252672704 A 80 80 0 0 1 322.504622962932 283.8094609392559 Z" fill="#50E3C2" stroke="none"/><path d="M 124.9295253889423 308.3213201202165 A 138 138 0 0 1 262.02749249917684 112.52513166333912" fill="none" stroke="#E91E63" stroke-width="2"/><line x1="250" y1="250" x2="32.486131111203974" y2="351.42838281776784" stroke="#000000" stroke-width="2"/><circle cx="32.486131111203974" cy="351.42838281776784" r="4" fill="#000000"/><text x="12.5473597963977" y="360.7259845760632" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text><line x1="250" y1="250" x2="29.07883517141437" y2="156.22452916257419" stroke="#000000" stroke-width="2"/><circle cx="29.07883517141437" cy="156.22452916257419" r="4" fill="#000000"/><text x="8.827728395460667" y="147.62844433581017" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><line x1="250" y1="250" x2="270.91737825943795" y2="10.913272457981066" stroke="#000000" stroke-width="2"/><circle cx="270.91737825943795" cy="10.913272457981066" r="4" fill="#000000"/><text x="272.8348045998864" y="-11.003010900037339" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><line x1="250" y1="250" x2="472.52412509602897" y2="160.09441758018113" stroke="#000000" stroke-width="2"/><circle cx="472.52412509602897" cy="160.09441758018113" r="4" fill="#000000"/><text x="492.9221698964983" y="151.85307252503105" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text><line x1="250" y1="250" x2="467.51386888879597" y2="351.4283828177679" stroke="#000000" stroke-width="2"/><circle cx="467.51386888879597" cy="351.4283828177679" r="4" fill="#000000"/><text x="487.4526402036023" y="360.72598457606324" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">E</text><text x="142.01644892310975" y="251.88485989522658" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">48°</text><text x="194.37588790971415" y="157.42593152417186" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">72°</text><text x="314.2408609691449" y="163.18345905334053" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">63°</text><text x="357.9629910973602" y="252.8271104172503" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">47°</text><text x="122.2122810909173" y="160.52207592923676" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">120°</text><circle cx="250" cy="250" r="6" fill="#000000"/><text x="250" y="275" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">O</text></svg>"`;

exports[`Widget: radially-constrained-angle-diagram should produce consistent output for example #5 1`] = `"<svg width="560" height="462" viewBox="-20 -22 560 462" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif"><path d="M 260 210 L 196.97230380721868 198.88651662931647 A 64 64 0 0 1 238.1107108271572 149.85967226970186 Z" fill="#f1c40f" stroke="none"/><path d="M 260 210 L 238.1107108271572 149.85967226970186 A 64 64 0 0 1 292 154.57437415779594 Z" fill="#2ecc71" stroke="none"/><path d="M 260 210 L 292 154.57437415779594 A 64 64 0 0 1 323.0276961927813 198.88651662931645 Z" fill="#9b59b6" stroke="none"/><path d="M 139.85345413251062 188.8149223246345 A 122 122 0 0 1 321 104.34490073829849" fill="none" stroke="#3498db" stroke-width="2"/><path d="M 96.52191299997347 181.17440250728959 A 166 166 0 0 1 423.47808700002656 181.17440250728956" fill="none" stroke="#E91E63" stroke-width="2"/><line x1="260" y1="210" x2="70.91691142165607" y2="176.6595498879494" stroke="#000000" stroke-width="2"/><circle cx="70.91691142165607" cy="176.6595498879494" r="4" fill="#000000"/><text x="49.25114085538749" y="172.8392899792769" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">B</text><line x1="260" y1="210" x2="194.3321324814716" y2="29.579016809105582" stroke="#000000" stroke-width="2"/><circle cx="194.3321324814716" cy="29.579016809105582" r="4" fill="#000000"/><text x="186.80768932830688" y="8.905779151815608" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">C</text><line x1="260" y1="210" x2="356" y2="43.723122473387775" stroke="#000000" stroke-width="2"/><circle cx="356" cy="43.723122473387775" r="4" fill="#000000"/><text x="367" y="24.670563590130143" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">D</text><line x1="260" y1="210" x2="449.0830885783439" y2="176.65954988794937" stroke="#000000" stroke-width="2"/><circle cx="449.0830885783439" cy="176.65954988794937" r="4" fill="#000000"/><text x="470.74885914461254" y="172.8392899792769" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">E</text><text x="189.52391123305404" y="150.86353990883836" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">60°</text><text x="268.01832833278456" y="118.35008777555942" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">50°</text><text x="335.36198807458726" y="157.23096785570377" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">50°</text><text x="187.23882655075238" y="89.45618374097948" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">110°</text><text x="260" y="24" font-size="18" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">160°</text><circle cx="260" cy="210" r="6" fill="#000000"/><text x="260" y="235" font-size="20" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#000000">A</text></svg>"`;
